integrationTests/chainSimulator/bridge/esdtSafe_test.go

```diff
@@ -0,0 +1,518 @@
+	esdtSafeAddrShard := chainSim.GetShardForAddress(cs, esdtSafeAddr)
+
+	wallets := generateAccountsAndTokens(t, cs)
+	amountToTransfer := big.NewInt(1)
+
+	// transfer main chain -> sovereign chain
+	// token originated from main chain
+	// expecting that tokens to be saved in esdt-safe contract after deposit
+	for account, token := range wallets {
+		// deposit 1 token, expect to be saved in contract and supply burned is 0
+		depositToken := chainSim.ArgsDepositToken{
+			Identifier: token.Identifier,
+			Nonce:      token.Nonce,
+			Amount:     amountToTransfer,
+			Type:       token.Type,
+		}
+		txResult := deposit(t, cs, account.Wallet.Bytes, &account.Nonce, bridgeData.ESDTSafeAddress, []chainSim.ArgsDepositToken{depositToken}, account.Wallet.Bytes)
+		chainSim.RequireSuccessfulTransaction(t, txResult)
+		chainSim.RequireAccountHasToken(t, cs, getTokenIdentifier(depositToken), account.Wallet.Bech32, big.NewInt(0).Sub(token.Amount, amountToTransfer))
+		waitIfCrossShardProcessing(cs, chainSim.GetShardForAddress(cs, account.Wallet.Bech32), esdtSafeAddrShard)
+		chainSim.RequireAccountHasToken(t, cs, getTokenIdentifier(depositToken), esdtSafeAddr, amountToTransfer)
+
+		_ = cs.GenerateBlocks(1)
+		tokenSupply, err := cs.GetNodeHandler(esdtSafeAddrShard).GetFacadeHandler().GetTokenSupply(getTokenIdentifier(depositToken))
+		require.Nil(t, err)
+		require.NotNil(t, tokenSupply)
+		require.Equal(t, "0", tokenSupply.Burned) // nothing burned
+	}
+
+	// transfer sovereign chain -> main chain
+	// token originated from main chain
+	// expecting that tokens will be transferred from contract balance after executeOperation
+	for account, token := range wallets {
+		executeToken := chainSim.ArgsDepositToken{
+			Identifier: token.Identifier,
+			Nonce:      token.Nonce,
+			Amount:     amountToTransfer,
+			Type:       token.Type,
+		}
+		// execute operations received from sovereign chain
+		// expecting the token to be transferred from esdt-safe contract to wallet address
+		txResult := executeOperation(t, cs, bridgeData.OwnerAccount.Wallet, account.Wallet.Bytes, &bridgeData.OwnerAccount.Nonce, bridgeData.ESDTSafeAddress, []chainSim.ArgsDepositToken{executeToken}, account.Wallet.Bytes, nil)
+		chainSim.RequireSuccessfulTransaction(t, txResult)
+		chainSim.RequireAccountHasToken(t, cs, getTokenIdentifier(executeToken), esdtSafeAddr, big.NewInt(0))
+		waitIfCrossShardProcessing(cs, esdtSafeAddrShard, chainSim.GetShardForAddress(cs, account.Wallet.Bech32))
+		chainSim.RequireAccountHasToken(t, cs, getTokenIdentifier(executeToken), account.Wallet.Bech32, token.Amount) // token original amount
+	}
+}
+
+func waitIfCrossShardProcessing(cs chainSim.ChainSimulator, senderShard uint32, receivedShard uint32) {
+	if senderShard != receivedShard {
+		_ = cs.GenerateBlocks(3)
+	}
+}
+
+func generateAccountsAndTokens(
+	t *testing.T,
+	cs chainSim.ChainSimulator,
+) map[*chainSim.Account]chainSim.ArgsDepositToken {
```

Comment:
map[*obj]?
That's really problematic, not ok.

Reply:
I was also afraid of mapping objects, but in this case is safe since your fields are comparable.

>If your obj is a struct, you must ensure that all fields within the struct are also comparable. If the struct contains a field like a slice, map, or a function, it cannot be used as a map key.

>Conclusion:
Using map[obj]obj in Go is perfectly valid as long as:

>- The key type is comparable.
>- You are aware of how the struct fields are used for equality and hashing.
>- Be cautious with structs that can be modified, as this could lead to unexpected behavior.